akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala

```diff
@@ -633,8 +639,8 @@ private[stream] final class GraphInterpreter(
 
   private[stream] def pull(connection: Int): Unit = {
     val currentState = portStates(connection)
+    portStates(connection) = currentState ^ PullStartFlip
     if ((currentState & OutClosed) == 0) {
-      portStates(connection) = currentState ^ PullStartFlip
```

Comment:
yeah, it is one of those visibility bugs :) Took a while to figure it out, even though I had the event trace of the failure. Btw, the reason why it was undetected is because it was usually happening like this:

 - since completion is usually reported early, the upstream boundary basically always finished with a push-then-complete
 - once the push and the complete was enqueued, the downstream pull could only arrive after the complete
 - this means, that push arrived first, resulting in a swallowed pull from buffer (because buffer is detached)
 - then complete arrived immediately, impossible to overtake by a downstream pull (because FIFO)
 - pull arrived after complete: either being ignored, or isHoldingUpstream returning false already

The reorder made it possible to observe the sequence of:
 - push-complete enqueued
 - pull enqueued
 - push arrives
 - pull arrives (reordered wrt complete)
 
and the rest of it is clear now. The nice thing about this fuzzer is that while we can still rely on the nice fairness properties of the FIFO queue, this fuzzer forces us to **not** rely it on being FIFO for the actual logics.